pyat/test/test_collective.py

```diff
@@ -0,0 +1,131 @@
+import at
+import numpy
+import pytest
+from numpy.testing import assert_allclose as assert_close
+from at.collective import Wake, WakeElement, ResonatorElement
+from at.collective import WakeComponent, ResWallElement
+from at.collective import add_beamloading, remove_beamloading, BLMode
+
+
+_issorted = lambda a: numpy.all(a[:-1] <= a[1:])
+
+
+def test_fillpattern(hmba_lattice):
+    ring = hmba_lattice.radiation_on(copy=True)
+    nbunch = 16
+    current = 0.2
+    ring.set_fillpattern(nbunch)
+    ring.beam_current = current
+    assert ring.fillpattern.shape == (ring.harmonic_number, )
+    assert numpy.sum(ring.fillpattern) == 1
+    assert ring.bunch_list.shape == (nbunch, )
+    assert ring.nbunch == nbunch
+    assert ring.bunch_currents.shape == (nbunch, )
+    assert numpy.sum(ring.bunch_currents) == current
+    
+
+def test_build_srange():
+    bspacing = 10
+    wlength = 20
+    bsize = 0.1
+    steps = 1.0e-3
+    stepl = 1.0e-2
+    srange = Wake.build_srange(-bsize, bsize, steps, stepl, bspacing, wlength)
+    assert len(numpy.unique(srange)) == len(srange)
+    assert _issorted(srange)
+    assert_close(max(srange), wlength + bsize - bspacing - stepl)
+    assert min(srange) == -bsize
+    assert_close(min(numpy.diff(srange)) ,steps)
+    assert_close(min(numpy.diff(srange[srange>bsize])) ,stepl)
+    
+                  
+def test_wake_object():
+    srange = [0, 1, 1]
+    with pytest.raises(Exception):
+        Wake(srange)
+    srange = Wake.build_srange(-0.36, 0.36, 1.0e-5, 1.0e-2, 844, 844)
+    with pytest.raises(Exception):
+        Wake.resistive_wall(srange, Wake.WakeComponent.X, 1.0, 1.0e-3, 1.0)
+    srange = Wake.build_srange(0.0, 0.36, 1.0e-5, 1.0e-2, 844, 844) 
+    long_res =  Wake.long_resonator(srange, 1.0e9, 1.0, 1.0e3, 1.0)
+    assert 1.0e-24 in long_res.srange
+    assert long_res.Z.shape == long_res.srange.shape
+    
+    
+def test_wake_element(hmba_lattice):
+    ring = hmba_lattice.radiation_on(copy=True)
+    srange = Wake.build_srange(0.0, 0.36, 1.0e-5, 1.0e-2, 844, 844) 
+    long_res =  Wake.long_resonator(srange, 1.0e9, 1.0, 1.0e3, 1.0) 
+    welem = WakeElement('WELEM', ring, long_res)
+    assert welem.TurnHistory.shape == (ring.nbunch*welem.Nslice*welem.Nturns, 4)
+    ring.append(welem)
+    ring.set_fillpattern(16)
+    assert welem.TurnHistory.shape == (ring.nbunch*welem.Nslice*welem.Nturns, 4)
+    ring.disable_6d(at.Collective)
+    assert welem.PassMethod == 'IdentityPass'
+    
+    
+def test_resonator_element(hmba_lattice):
+    ring = hmba_lattice.radiation_on(copy=True)
+    srange = Wake.build_srange(0.0, 0.36, 1.0e-5, 1.0e-2, 844, 844)
+    welem = ResonatorElement('WELEM', ring, srange, WakeComponent.Z, 1.0e9, 1.0, 1.0e3)  
+    assert welem.ResFrequency == 1.0e9
+    assert welem.Qfactor == 1
+    assert welem.Rshunt == 1.0e3 
+    welem.ResFrequency += 100
+    assert welem.ResFrequency == 1.0e9 + 100
+    
```

Comment:
I think it may be useful to test the values of the first 2 or 3 indices. Only because we we have some minus sign manipulation between s and t.